![](https://www.dii.uchile.cl/wp-content/uploads/2021/06/Magi%CC%81ster-en-Ciencia-de-Datos.png)

**MDS7202: Laboratorio de Programación Científica para Ciencia de Datos**

### 👨‍🏫👩‍🏫 Cuerpo Docente:

- Profesores: Diego Cortez, Gabriel Iturra
- Auxiliares: Melanie Peña, Valentina Rojas
- Ayudantes: Nicolás Cabello, Cristopher Urbina

### 👨‍💻👩‍💻 Estudiantes:
- Estudiante n°1: Josefa Anselmo.
- Estudiante n°2: Tamara Carrasco.

_Por favor, lean detalladamente las instrucciones de la tarea antes de empezar a escribir._

---

---

## 📖 Enunciado 

<div align="center">
    <img src='https://github.com/MDS7202/MDS7202/blob/main/recursos/2025-01/proyecto/proyecto.png?raw=true' style="border-radius: 12px"> 
</div>

En el competitivo universo de las bebidas gaseosas, la empresa **SodAI Drinks 🥤** ha logrado destacarse por su creatividad, diversidad de productos y enfoque centrado en el cliente. Ofrece una extensa gama de bebidas carbonatadas que abarca distintos segmentos del mercado: desde productos premium en presentaciones sofisticadas, hasta gaseosas accesibles para el consumo masivo, disponibles en diversos tamaños y tipos de envases. 

La compañía opera en múltiples regiones y zonas, sirviendo a una variedad de puntos de venta que incluyen desde tiendas de conveniencia y minimarkets hasta el canal frío tradicional. Cada tipo de cliente tiene sus particularidades: algunos reciben entregas hasta 4 veces por semana, mientras que otros son visitados por la fuerza de ventas solo una vez semanalmente. Esta diversidad de perfiles representa tanto una oportunidad como un desafío comercial: ¿cómo saber qué productos tienen más chances de ser comprados por cada cliente en un momento dado?

Con el objetivo de aumentar la facturación de forma inteligente y mejorar la eficiencia de su estrategia de ventas, **SodAI Drinks** decide crear una nueva célula interna de innovación: el equipo **Deep Drinkers 🤖**, cuyo propósito es aplicar ciencia de datos para anticiparse a las necesidades del cliente y potenciar el negocio desde una perspectiva basada en información.

El corazón de esta iniciativa es el desarrollo de un sistema predictivo personalizado para cada cliente. Para ello, **Deep Drinkers** convoca a un equipo de Data Scientists y especialistas en *machine learning* con una misión clara: construir un modelo predictivo que, cada semana, pueda estimar la probabilidad de compra de cada producto del portafolio para cada cliente activo.

El modelo deberá tener en cuenta múltiples factores, incluyendo:
- **Tipo de cliente**, ej. "TIENDA DE CONVENIENCIA", "MINIMARKET".
- **Frecuencia de entregas y visitas**, indicadores del nivel de actividad comercial.
- **Ubicación geográfica** (por región y zona).
- **Preferencias históricas de consumo**, inferidas por patrones de compra anteriores.
- **Características del producto**, como marca, categoría, segmento, tipo de envase y tamaño

El objetivo final es que, **cada semana**, se genere una tabla de productos priorizados: para cada cliente, un listado de productos ordenado por su probabilidad estimada de compra. Esta información será enviada al equipo comercial, que podrá usarla en call center, para incrementar las chances de concretar ventas al ofrecer justo lo que el cliente probablemente quiere comprar.

Este proyecto representa un cambio de paradigma en la forma en que **SodAI Drinks** gestiona su fuerza de ventas: de un enfoque reactivo y generalista, a uno proactivo, basado en datos y profundamente personalizado. Así, la empresa no solo espera aumentar su rentabilidad, sino también construir relaciones más sólidas con sus clientes, ofreciéndoles recomendaciones más relevantes y oportunas.

Para lograr lo anterior, el equipo **Deep Drinkers** contará con los siguientes conjuntos de datos, junto a sus respectivos atributos:

- **Datos transaccionales** (`transacciones.parquet`): contiene el historial de compras realizadas por los clientes.
	- `customer_id`: identificador único del cliente que realizó la compra.
	- `product_id`: identificador único del producto comprado.
	- `purchase_date`: fecha en que se realizó la transacción.
	- `order_id`: identificar de la orden de su pedido.
	- `items`: cantidad de bultos comprados por cliente en aquella transacción.

- **Datos de clientes** (`clientes.parquet`): incluye las características de cada cliente.
	- `customer_id`: identificador único del cliente.
	- `region_id`: identificador de la región geográfica donde se encuentra el cliente.
	- `customer_type`: tipo de cliente según el canal comercial, por ejemplo, “TIENDA DE CONVENIENCIA”.
	- `Y`: coordenada geográfica de latitud.
	- `X`: coordenada geográfica de longitud.
	- `num_deliver_per_week`: cantidad de entregas semanales que recibe el cliente.
	- `num_visit_per_week`: frecuencia de visitas de la fuerza de ventas por semana.

- **Datos de productos** (`productos.parquet`): describe las características de los productos del portafolio.
	- `product_id`: identificador único del producto.
	- `brand`: marca comercial del producto.
	- `category`: categoría general del producto, como “BEBIDAS CARBONATADAS”.
	- `sub_category`: subcategoría dentro de la categoría principal, por ejemplo, “GASEOSAS”.
	- `segment`: segmento de mercado al que pertenece el producto, como “PREMIUM”.
	- `package`: tipo de envase del producto.
	- `size`: tamaño del producto en litros.

## 📚 Reglas

<center>
<img src="https://media1.tenor.com/m/0Qtv_cQ4ITsAAAAd/necohaus-grey-name.gif" width="450">



El proyecto consta de **dos entregas parciales** y una **entrega final** en donde la primera entrega la idea es poder reflejar lo aprendido durante la primera mitad del curso, que será sobre los contenidos relacionados a *machine learning*, la segunda será sobre los contenidos de la segunda mitad del curso relacionados a *MLOps* y por último la entrega final constará de dos partes, donde la primera será relacionada con experimentación sobre nuevos datasets que serán disponibilizados durante las últimas semanas del curso de manera incremental y una segunda parte que será el informe final escrito que deberá explicar el desarrollo del proyecto completo, como tambien los resultados y análisis de los experimentos realizados sobre los datasets incrementales. La idea es que todo el código esté desarrollado durante las primeras dos entregas y luego en la entrega final sólo se ejecute el código sobre nuevos conjuntos de datos.

La idea de generar el proyecto por etapas es poder aliviar la carga de trabajo en las últimas semanas del semestre donde sabemos que están muy cargado con entregas, pruebas y exámenes de otros ramos, y así garantizamos que habiendo la desarrollado las dos primeras entregas parciales, tendrán el grueso del proyecto listo para luego experimentar y documentar.

---
### **Fechas de entrega**
- **Entrega parcial 1**: 12 de Septiembre
- **Entrega parcial 2**: Por definir
- **Entrega final**: Por definir

---

### **Requisitos del proyecto**
- **Grupos**: Formar equipos de **2 personas**. No se aceptarán trabajos individuales o grupos con más integrantes.
- **Consultas**: Cualquier duda fuera del horario de clases debe ser planteada en el foro correspondiente. Los mensajes enviados al equipo docente serán respondidos únicamente por este medio. Por favor, revisen las respuestas anteriores en el foro antes de realizar nuevas consultas.
- **Plagio**: La copia o reutilización no autorizada de trabajos de otros grupos está **estrictamente prohibida**. El incumplimiento de esta norma implicará la anulación inmediata del proyecto y una posible sanción académica.
- **Material permitido**: Pueden usar cualquier material del curso, ya sea notas, lecturas, códigos, o referencias proporcionadas por los docentes, que consideren útil para el desarrollo del proyecto.

---

### **Entregables y etapas**

#### **1. Entrega Parcial 1**  
- Dispondrán de los archivos de datos **productos.parquet**, **clientes.parquet** y **transacciones.parquet** para el modelamiento inicial.  
- Utilizarán estos archivos para desarrollar lo solicitado para la entrega 1. 
- En esta etapa, se espera que apliquen todos los conocimientos aprendidos durante la primera parte del curso relacionados con *machine learning*.
- **Informe**: No se exige un avance del informe en esta etapa, sólo un notebook con su desarrollo actual, pero se **recomienda comenzar** a redactar el informe final en paralelo para disminuir la carga académica en las etapas posteriores.  

#### **2. Entrega Parcial 2**  
- En esta entrega, deberán aplicar los conocimientos aprendidos durante la segunda mitad del curso sobre *MLOps*  
- Se espera que implementen estos conocimientos para desplegar su modelo elegido en la primera entrega y crear *pipelines* automatizados que simulen un entorno productivo.
- **Informe**: similar a la primera etapa, no se exige un avance del informe, pero se **recomienda avanzar con su redacción** para evitar una acumulación de trabajo en la etapa final.  

#### **3. Entrega Final**  
- En la entrega final, deberán realizar dos etapas:
	- La primera etapa es sobre experimentación utilizando datasets incrementales que se irán disponibilizando de manera parcial, para que vayan generando predicciones con su modelo ya desplegado. El objetivo de esta etapa es poder testear su solución *end-to-end* y que vayan analizando los resultados obtenidos a medida que se van agregando más datos.
	- La segunda etapa consiste en redactar un informe final que deberá explicar el desarrollo completo de tu proyecto y un análisis profundo de sus resultados de experimentación. Este informe debera incluir a lo menos las siguientes secciones:
		- Análisis exploratorio de datos  
		- Metodología aplicada  
		- Selección y entrenamiento de modelos  
		- Evaluación de resultados  
		- Optimización de modelos
		- Interpretabilidad
		- Re-entrenamiento
		- Tracking con MLFlow
		- Creación de la aplicación web con Gradio y FastAPI

Es **altamente recomendable** ir redactando el informe en paralelo al desarrollo de los modelos para garantizar que toda la información relevante quede documentada adecuadamente.  

### Nota Final

La calificación final de su proyecto se calculará utilizando la siguiente ponderación: 

$$Nota Final = 0.30 * EntregaParcial1 + 0.40 * EntregaParcial2 + 0.30 * EntregaFinal$$

---

### **Instrucciones importantes**

1. **Formato del informe**:  
   - El informe debe estar integrado dentro de un **Jupyter Notebook**. No es necesario subirlo a una plataforma externa, pero debe cumplir con los siguientes requisitos:  
     - Estructura clara y ordenada.  
     - Código acompañado de explicaciones detalladas.  
     - Resultados presentados de forma visual y analítica.  

2. **Descuento por informes deficientes**:  
   - Cualquier sección del informe que no tenga una explicación adecuada o no respete el formato será penalizada con un descuento en la nota. Esto incluye código sin comentarios o análisis que no sean coherentes con los resultados presentados.
   - Comentarios sin formatear de ChatGPT o herramientas similares serán penalizados (e.g: "Inserta tu modelo acá", etc.)

# 📬 Entrega Parcial 1 (30% del Proyecto)

### 📪 Fecha de Entrega: 12 de Septiembre

## 📌 Abstract [0.25 puntos]

<center>
<img src="https://i.redd.it/h5ptnsyabqvd1.gif" width="400" height="300">

En esta sección, deben redactar un Abstract claro y conciso para su proyecto. El Abstract debe responder a las siguientes preguntas clave:

- **Descripción del problema**: ¿Cuál es el objetivo del proyecto? ¿Qué se intenta predecir o analizar?
- **Datos de entrada**: ¿Qué datos tienen disponibles? ¿Cuáles son sus principales características?
- **Métrica de evaluación**: ¿Cómo medirán el desempeño de sus modelos? Expliquen por qué eligieron esta métrica basándose en el análisis exploratorio de los datos.
- **Modelos y transformaciones**: ¿Qué modelos utilizarán y por qué? ¿Qué transformaciones o preprocesamientos aplicaron a los datos?
- **Resultados generales**: ¿El modelo final cumplió con los objetivos del proyecto? ¿Cuáles fueron las conclusiones más importantes?

**Importante**: Escriban esto despues de haber resuelto el resto de la tarea.

> [Escriba aquí su Abstract]

## 📌 Pre-procesamiento [0.5 puntos]

<center>
<img src="https://media0.giphy.com/media/10zsjaH4g0GgmY/giphy.gif?cid=6c09b9523xtlunksc9amikw09zk1bmiqwjqnt70ae82rk877&ep=v1_gifs_search&rid=giphy.gif&ct=g" width="400" height="300">

Tal como en muchos otros problemas de negocio, los datos probablemente deben ser pre procesados antes de aplicar cualquier técnica de analítica. Bajo esa premisa, en esta sección deben desarrollar código que les permita **preparar los datos** de tal forma que les permita resolver el problema planteado. Para esto, pueden aplicar procesamientos como:

- Transformaciones de tipo de dato (str, int, etc)
- Cruce de información
- Eliminación de duplicados
- Filtros de fila y/o columnas

*Hint: ¿Qué forma debería tener la data para resolver un problema de aprendizaje supervisado?*

Todo proceso llevado a cabo debe estar bien documentado y justificado en el informe, explicando el por qué se decidió realizar en funcion de los datos presentados y los objetivos planteados del proyecto.

In [26]:
# Desarrolle aquí su código

# librerías para EDA

import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px

# Carga de archivos

clientes = pd.read_parquet("clientes.parquet")
productos = pd.read_parquet("productos.parquet")
transacciones = pd.read_parquet("transacciones.parquet")

**Aplicamos .head() para ver cómo es cada tabla gráficamente**

In [27]:
print("="*10 + " Primeros 3 registros del dataset Clientes "+ "="*10)
clientes.head(3)

========== Primeros 3 registros del dataset Clientes ==========


customer_id  region_id  zone_id customer_type          Y           X  \
10705       256017         80     5148     ABARROTES -46.474800 -108.045140   
10706       255780         80     5148     ABARROTES -46.520282 -107.961052   
10707       254655         80     5148     ABARROTES -46.537640 -107.909280   

       num_deliver_per_week  num_visit_per_week  
10705                     1                   1  
10706                     1                   1  
10707                     1                   1

In [28]:
print("="*10 + " Primeros 3 registros del dataset Productos "+ "="*10)
productos.head(3)

========== Primeros 3 registros del dataset Productos ==========


product_id     brand              category sub_category  segment  package  \
0       34092  Brand 31  BEBIDAS CARBONATADAS     GASEOSAS  PREMIUM  BOTELLA   
1       57290  Brand 31  BEBIDAS CARBONATADAS     GASEOSAS  PREMIUM  BOTELLA   
2       56714  Brand 31  BEBIDAS CARBONATADAS     GASEOSAS  PREMIUM  BOTELLA   

   size  
0  0.66  
1  0.66  
2  0.66

In [29]:
print("="*10 + " Primeros 3 registros del dataset Transacciones "+ "="*10)
transacciones.head(3)

========== Primeros 3 registros del dataset Transacciones ==========


customer_id  product_id  order_id purchase_date     items
124        61353       61364    411145    2024-04-27 -0.333333
127        61899        1370    411156    2024-04-27  3.666667
344       154077       30500    417911    2024-04-29  2.333333

**Revisamos los .shape (dimensiones) de las tablas**

In [30]:
# dimensiones de los dataset

print("="*10 + " Dimensiones de los datasets " + "="*10)

print("Clientes: ", clientes.shape)

print("Productos: ", productos.shape)

print("Transacciones: ", transacciones.shape)

========== Dimensiones de los datasets ==========
Clientes:  (1569, 8)
Productos:  (971, 7)
Transacciones:  (254936, 5)


**Aplicamos .tolist() para poder tener identificados los nombres de las variables que usamos en cada tabla**

In [31]:
# Nombres de columnas identificadas de cada archivo

client_columns = clientes.columns.tolist()

products_columns = productos.columns.tolist()

transactions_columns = transacciones.columns.tolist()

print("columnas del archivo cliente: ", client_columns)

print("Columnas del archivo productos: ", products_columns)

print("Columnas del archivo transacciones", transactions_columns)


columnas del archivo cliente:  ['customer_id', 'region_id', 'zone_id', 'customer_type', 'Y', 'X', 'num_deliver_per_week', 'num_visit_per_week']
Columnas del archivo productos:  ['product_id', 'brand', 'category', 'sub_category', 'segment', 'package', 'size']
Columnas del archivo transacciones ['customer_id', 'product_id', 'order_id', 'purchase_date', 'items']


**Aquí dejo los tipos de dato que tiene cada tabla con .dtypes**

In [32]:
# Tipo de datos identificados en las columnas
print("="*30 + "Tipo de valores en las columnas (Cliente) "+ "="*30)
print(clientes.dtypes)


print("="*30 + "Tipo de valores en las columnas (productos) "+ "="*30)
print(productos.dtypes)


print("="*30 + "Tipo de valores en las columnas (transacciones) "+ "="*30)
print(transacciones.dtypes)
print("="*80)

==============================Tipo de valores en las columnas (Cliente) ==============================
customer_id               int64
region_id                 int64
zone_id                   int64
customer_type            object
Y                       float64
X                       float64
num_deliver_per_week      int64
num_visit_per_week        int64
dtype: object
==============================Tipo de valores en las columnas (productos) ==============================
product_id        int64
brand            object
category         object
sub_category     object
segment          object
package          object
size            float64
dtype: object
==============================Tipo de valores en las columnas (transacciones) ==============================
customer_id               int64
product_id                int64
order_id                  int64
purchase_date    datetime64[ns]
items                   float64
dtype: object


**Las columnas que funcionana como un id ("customer_id", "region_id" y "zone_id", product_id, "order_id") en las 3 tablas ahora estan como número y no deberiamos tratarlas porque estas son más bien identificadores nominales o sea que deberian ser texto en el fondo. En las tablas, las varibales customer_type, brand, category, sub_category, segmente y package estan como object, pero para orden y para que sea más explicita la convertiremos en category directamete. Además, esto disminuye el costo computacional del código porque si usamos clase object, pandas considera cada valor como un puntero a un objeto de Python mientras qeu al estar como category se crea como "un diccionario" interno y eso pesa menos.**

In [33]:
# Modificamos dtypes pues habían ciertas incongruencias con los dtypes actuales

# Clientes
clientes = clientes.astype({
    "customer_id": "string",
    "region_id": "string",
    "zone_id": "string",
    "customer_type": "category"
})

# Productos
productos = productos.astype({
    "product_id": "string", # Podría considerarse como category??
    "brand": "category",
    "category": "category",
    "sub_category": "category",
    "segment": "category",
    "package": "category"
})

# Transacciones
transacciones = transacciones.astype({
    "customer_id": "string",
    "product_id": "string",
    "order_id": "string"
})

In [34]:
# Confirmo cambios
print("="*30 + "Tipo de valores en las columnas (Cliente) "+ "="*30)
print(clientes.dtypes)


print("="*30 + "Tipo de valores en las columnas (productos) "+ "="*30)
print(productos.dtypes)


print("="*30 + "Tipo de valores en las columnas (transacciones) "+ "="*30)
print(transacciones.dtypes)
print("="*80)

==============================Tipo de valores en las columnas (Cliente) ==============================
customer_id             string[python]
region_id               string[python]
zone_id                 string[python]
customer_type                 category
Y                              float64
X                              float64
num_deliver_per_week             int64
num_visit_per_week               int64
dtype: object
==============================Tipo de valores en las columnas (productos) ==============================
product_id      string[python]
brand                 category
category              category
sub_category          category
segment               category
package               category
size                   float64
dtype: object
==============================Tipo de valores en las columnas (transacciones) ==============================
customer_id      string[python]
product_id       string[python]
order_id         string[python]
purchase_date    datetime64[ns

**Aplico estadísticas descriptivas básicas ahora que tengo correcto el tipo de dato -> solo se aplica a las columnas numéricas en este caso**

In [35]:
# Aplico la función definida sobre el dataset clientes
print("="*10 + "Estadística descriptiva clientes (numérico)" + "="*10)
print(clientes.describe())
print("="*10 + "Estadística descriptiva productos (numérico)" + "="*10)
print(productos.describe())
print("="*10 + "Estadística descriptiva transacciones (numérico)" + "="*10)
print(transacciones.describe())

==========Estadística descriptiva clientes (numérico)==========
                 Y            X  num_deliver_per_week  num_visit_per_week
count  1569.000000  1568.000000           1569.000000              1569.0
mean    -46.700549  -107.743548              3.104525                 1.0
std       2.700406     2.693112              0.515610                 0.0
min    -109.002508  -108.620920              1.000000                 1.0
25%     -46.586122  -107.934477              3.000000                 1.0
50%     -46.555514  -107.894765              3.000000                 1.0
75%     -46.513968  -107.862156              3.000000                 1.0
max     -46.161950   -46.442785              6.000000                 1.0
==========Estadística descriptiva productos (numérico)==========
             size
count  971.000000
mean     1.420525
std      3.897778
min      0.125000
25%      0.310000
50%      0.500000
75%      1.000000
max     20.000000
==========Estadística descriptiva transacci

**Aplico para string y separado para category**

In [36]:
print("="*10 + "Estadística descriptiva clientes (string)" + "="*10)
print(clientes.describe(include="string"))
print("="*10 + "Estadística descriptiva productos (string)" + "="*10)
print(productos.describe(include="string"))
print("="*10 + "Estadística descriptiva transacciones (string)" + "="*10)
print(transacciones.describe(include="string"))

==========Estadística descriptiva clientes (string)==========
       customer_id region_id zone_id
count         1569      1569    1569
unique        1569         1       1
top         255780        80    5148
freq             1      1569    1569
==========Estadística descriptiva productos (string)==========
       product_id
count         971
unique        971
top         50254
freq            1
==========Estadística descriptiva transacciones (string)==========
       customer_id product_id order_id
count       254936     254936   254936
unique        1490        114    64600
top         183105       1370    23039
freq          1619      37710       47


**Transacciones no tiene una varible tipo category**

In [37]:
print("="*10 + "Estadística descriptiva clientes (category)" + "="*10)
print(clientes.describe(include="category"))
print("="*10 + "Estadística descriptiva productos (category)" + "="*10)
print(productos.describe(include="category"))

==========Estadística descriptiva clientes (category)==========
       customer_type
count           1569
unique             7
top        ABARROTES
freq            1160
==========Estadística descriptiva productos (category)==========
          brand              category sub_category  segment  package
count       971                   971          971      971      971
unique       61                     2            3        4        4
top     Brand 7  BEBIDAS CARBONATADAS     GASEOSAS  PREMIUM  BOTELLA
freq        104                   743          673      310      597


Las estadísticas que importan del dataset de ``clientes.parquet`` son de las columnas ``num_deliver_per_week``, ``num_visit_per_week`` y ``customer_type``

Acá se identifica que tal vez sea más conveniente transformar la columna fecha de compra -> que aparezcan 3 columnas extras (año, mes, día).

Las columnas cuyas estadísticas importan: ``items``: Se identifca un valor negativo, cuando esta columna representa a cantidad de bultos comprados por el cliente en aquella transacción.

**La variable purchase_date tiene otro tipo de dato (datetime64[ns]) por lo que la revisaremos por separado**

In [ ]:
# Tipo de dato
transacciones["purchase_date"].dtype

# Ver encabezado para saber como 
print(transacciones["purchase_date"].head(3))

# Reviso rango
print("Min:", transacciones["purchase_date"].min())
print("Max:", transacciones["purchase_date"].max())

# Veo nulos y únicos
print("Unique:", transacciones["purchase_date"].nunique())
print("Nulos:", transacciones["purchase_date"].isna().sum())

124   2024-04-27
127   2024-04-27
344   2024-04-29
349   2024-04-29
354   2024-04-26
359   2024-04-29
362   2024-04-29
363   2024-04-29
365   2024-04-28
367   2024-04-29
Name: purchase_date, dtype: datetime64[ns]
Min: 2024-01-01 00:00:00
Max: 2024-12-31 00:00:00
Unique: 366
Nulos: 0


**Si bien pareciera que solo se presenta en la columna el año, mes y día (en ese orden), queremos confirmar que no hayan registos de hora que en el fondo igual son soportados por lo que se muestra en la estadística básica de los mínimos y máximos.**
**Transacciones de un año**

In [39]:
from datetime import time # Para poder trabajar directamente con fechas/tiempo

print(transacciones["purchase_date"].dt.time.nunique())  # quiero saber si la parte del tiempo es el mismo en todas, o sea como 00:00:00 y así asegurarme que es como "por defecto" o si es que de hecho la columna captura esa información

1


In [40]:
# Agregaremos las variables de fecha específica (día, mes y año) para posibles futuros desarrollos
transacciones["year"] = transacciones["purchase_date"].dt.year
transacciones["month"] = transacciones["purchase_date"].dt.month
transacciones["day"] = transacciones["purchase_date"].dt.day

In [41]:
transacciones.head(2)

customer_id product_id order_id purchase_date     items  year  month  day
124       61353      61364   411145    2024-04-27 -0.333333  2024      4   27
127       61899       1370   411156    2024-04-27  3.666667  2024      4   27

**Vamos a dejar igual la variable de purchase_date**

In [42]:
transacciones.shape

(254936, 8)

**Reviso si hay valores nulos en las tablas con isna**

In [43]:
# Identificación de valores nulos

print(" Identificación de valores nulos por columna de cada dataset")

print("="*30 + " Dataset clientes "+ "="*30)
print(clientes.isna().sum())

print("="*30 + " Dataset productos "+ "="*30)
print(productos.isna().sum())

print("="*30 + " Dataset transacciones "+ "="*30)
print(transacciones.isna().sum())

 Identificación de valores nulos por columna de cada dataset
============================== Dataset clientes ==============================
customer_id             0
region_id               0
zone_id                 0
customer_type           0
Y                       0
X                       1
num_deliver_per_week    0
num_visit_per_week      0
dtype: int64
============================== Dataset productos ==============================
product_id      0
brand           0
category        0
sub_category    0
segment         0
package         0
size            0
dtype: int64
============================== Dataset transacciones ==============================
customer_id      0
product_id       0
order_id         0
purchase_date    0
items            0
year             0
month            0
day              0
dtype: int64


En el dataset de ``clientes.parquet``, hay un registro que no cuenta con coordenada geográfica X (longitud).

**Reviso los duplicados con duplicated**

In [44]:
# identificación de valores duplicados

print(" Identificación de registros duplicados por columna de cada dataset")
print("="*30 + " Dataset clientes "+ "="*30)
print(clientes.duplicated(subset=["customer_id"]).sum())

print("="*30 + " Dataset productos "+ "="*30)
print(productos.duplicated().sum())

print("="*30 + " Dataset transacciones "+ "="*30)
print(transacciones.duplicated().sum())

 Identificación de registros duplicados por columna de cada dataset
============================== Dataset clientes ==============================
0
============================== Dataset productos ==============================
0
============================== Dataset transacciones ==============================
885


Existen 885 transacciones duplicadas, las cuales tendrán que ser trabajadas también

In [45]:
# Eliminamos tuplas duplicadas

transacciones = transacciones.drop_duplicates(subset=["order_id", "product_id"], keep="first") # Defino que revise y bote duplicados basandose en la llave de alguna forma, en el fonod cada producto se asocia a una orden y el cliente esta implicito
print("Dimension del dataset después de eliminar duplicados: ", transacciones.shape) 
transacciones.head(2)

Dimension del dataset después de eliminar duplicados:  (243338, 8)


customer_id product_id order_id purchase_date     items  year  month  day
124       61353      61364   411145    2024-04-27 -0.333333  2024      4   27
127       61899       1370   411156    2024-04-27  3.666667  2024      4   27

# Cruce de información.
En este punto, es válido realizar un cruce de información usando las llaves primarias, tales como ``Customer ID`` del dataset Clientes y ``Product ID`` del dataset Productos. En el dataset de transacciones, las llaves mencionadas ahora son llaves foráneas y se tiene la llave primaria ``Order ID``.

Como el dataset Transacciones contiene esas 3 llaves, comenzamos el merge desde ahí: primero con clientes y luego con productos.

**Entonces hare los merge y después volveré a revisar los datos más o menos para saber cómo quedo y asegurarme de que quedo bien**

In [46]:
# Hago merge de los dataset en un solo dataframe -> df

df = transacciones.merge(clientes, on="customer_id", how="left") # how = left significa que usa solo las llaves del frame izquierdo, como un
                                                                # sql left outer join: preserva el orden de la llave -> deberia aparecer customer id como 
                                                                # primera columna

df = df.merge(productos, on="product_id", how="left") 

**Entonces ahora los 3 quedan como un gran data frame que convservan la información principalmente de las transacciones, esto debido a que eso es lo que "confirma" que hubo una compra. Puedo tener quizas algunos clientes que se eliminen en le merge o productos que se eliminen, pero debería primar el tema de las transacciones porque ahí queda registrada la compra por decir. Los otros son como "perfil" de cliente y "perfil" de objeto que se vende**

In [47]:
print("Dimensiones del dataset: ",df.shape)
df.head(2)

Dimensiones del dataset:  (243338, 21)


customer_id product_id order_id purchase_date     items  year  month  day  \
0       61353      61364   411145    2024-04-27 -0.333333  2024      4   27   
1       61899       1370   411156    2024-04-27  3.666667  2024      4   27   

  region_id zone_id  ...          Y           X  num_deliver_per_week  \
0        80    5148  ... -46.558718 -107.860564                     3   
1        80    5148  ... -46.557632 -107.896470                     3   

   num_visit_per_week     brand              category sub_category segment  \
0                   1  Brand 35  BEBIDAS CARBONATADAS     GASEOSAS  MEDIUM   
1                   1  Brand 34  BEBIDAS CARBONATADAS     GASEOSAS  MEDIUM   

   package size  
0  BOTELLA  1.0  
1  BOTELLA  1.0  

[2 rows x 21 columns]

In [48]:
print("Valores Nulos en dataset df")
print(df.isna().sum())

Valores Nulos en dataset df
customer_id             0
product_id              0
order_id                0
purchase_date           0
items                   0
year                    0
month                   0
day                     0
region_id               0
zone_id                 0
customer_type           0
Y                       0
X                       0
num_deliver_per_week    0
num_visit_per_week      0
brand                   0
category                0
sub_category            0
segment                 0
package                 0
size                    0
dtype: int64


Acá desapareció el único cliente que tenía una coordenada "X" como NaN. Por cómo se hizo el merge, capaz que el cliente no estaba en transacciones y ahora desapareció del df final.

**Ya entonces viendo las dimensiones y si es que hay nulos, puedo presumir que todas las transacciones fueron hechas por clientes que exitian y se vendieron items que estaban registrados. Además, podría ser que en el merge se perdio info de los items o de los clientes pero apra terminos de compras esta informacion no debiera ser relevante y por ella esta bien si se perdió pues no nos aporta concretamente algo ahora**

In [49]:
# Revisar si efectivamente hubieron registros de cliente y de producto que no se mergearon con la tabla transacciones pues no compraron/no figuraban en inventario

print("Clientes en clientes:", clientes["customer_id"].nunique())
print("Clientes en transacciones:", df["customer_id"].nunique())
print("Clientes faltantes:", len(set(clientes["customer_id"]) - set(df["customer_id"])))

print("Productos en productos:", productos["product_id"].nunique())
print("Productos en transacciones:", df["product_id"].nunique())
print("Productos faltantes:", len(set(productos["product_id"]) - set(df["product_id"])))

Clientes en clientes: 1569
Clientes en transacciones: 1490
Clientes faltantes: 79
Productos en productos: 971
Productos en transacciones: 114
Productos faltantes: 857


**Efectivamente entonces no todos los clientes y productos pasan al merge, pero esto se explica en lo que ya dijimos ants, o sea, no todos los clientes han comprado ni todos los productos se han transado**


Sobre la columna de la fecha: ya está en un buen formato, pero en el momento en que se requiera un análisis temporar: mes, semana, día... se pueden crear 
columnas auxiliares.


In [50]:
# valores duplicados en df nuevo mergeado

df.duplicated().sum()

np.int64(0)

**como se puede ver tenemos 885 transacciones duplicadas, como esto no es "posibble" teneoms que droppear las tuplas duplicadas pues solo nos estan duplicando, valga la redundancia, la información y no nos estan aportando más datos**

**Super importante no olvidar que el drop y las modificaciones se han hecho todo el tiempo considerando que una transacción equivale a 1 registro. Que la primary key es el order_id, por lo que solo puede haber 1 order_id x registro**

## 📌 EDA [0.5 puntos]

<center>
<img src="https://media3.giphy.com/media/v1.Y2lkPTc5MGI3NjExbHZ6aGdkd21tYTI3cW8zYWhyYW5wdGlyb2s3MmRzeTV0dzQ1NWlueiZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/3k1hJubTtOAKPKx4k3/giphy.gif" width="400" height="200">

En esta sección, se debe realizar un análisis exploratorio de los datos para comprender su estructura, detectar posibles problemas y obtener información relevante para el entrenamiento de los modelos. La idea es que puedan detectar **patrones en los datos** que les permitan resolver el problema con mayor facilidad.

Se deben responder preguntas a partir de lo que puedan visualizar/obtener, por ejemplo:

- Clientes y productos

    - ¿Cuántos clientes únicos hay en el dataset?

    - ¿Cuántos productos únicos se encuentran en los datos?

- Periodo y frecuencia

    - ¿De qué periodo es la información disponible?

    - ¿Cuál es la frecuencia de los registros (diaria, semanal, mensual, etc.)?

- Calidad de los datos

    - ¿Existen valores nulos en el dataset? ¿Cuántos? ¿Cómo se pueden tratar?

    - ¿Hay datos raros, como cantidades negativas o inconsistencias? Genere tests de validación para identificar estos problemas.

- Patrones de compra

    - ¿Cuántos productos compra en promedio cada cliente semana a semana?

    - ¿Cuántas transacciones ha realizado cada cliente?

    - ¿Cuál es el periodo de recompra promedio de cada SKU?

In [ ]:
# Desarrolle aquí su código

# clientes y productos

df["customer_id"].value_counts().shape # hay 1490 clientes únicos

df["product_id"].value_counts().shape # hay 114 productos únicos

In [ ]:
# sobre las fechas

fecha_inicio = df["purchase_date"].min()

fecha_final = df["purchase_date"].max()

print("Fecha inicio de observaciones: ", fecha_inicio)# los registros son desde el 1 de enero del año 2024?

print("Fecha final de observaciones: ", fecha_final )



In [ ]:
df = df.set_index("purchase_date") # esto ordena las filas del df según la fecha

frecuencia_diaria = df.resample('D').size().reset_index(name='frecuencia')
print(frecuencia_diaria)

In [ ]:
frecuencia_semanal = df.resample('W').size().reset_index(name='frecuencia')
print(frecuencia_semanal)

In [ ]:
frecuencia_mensual = df.resample("ME").size().reset_index(name="frecuencia") # M o ME para "month"
print(frecuencia_mensual)

De acá, los meses en donde se compran más productos son  los menes de fin de año (Noviembre-Diciembre) y principio de año (Enero). El mes con ventas más bajas es el mes de Mayo.
Sobre cantidades negativas, hay items que salen con signo negativo.


In [ ]:
# Columnas del dataset mergeado

cat_columns = column_type(df, printeo="no")[0]
binary_columns = column_type(df, printeo="no")[1]
num_columns = column_type(df, printeo="no")[2]
cat_nominal_columns = column_type(df, printeo="no")[3]
cat_ordinal_columns = column_type(df, printeo="no")[4]



In [ ]:
cat_columns

Comentario: Acá me di cuenta cuenta que hay variables categóricas nominales que fueron colocadas como ordinales Como customer_type y package, que son variables categóricas nominales (Aunque customer_type es debatible).
En cualquier caso, solo hay que modificar la función column_type, colocando las variables categoricas nominales a mano.

La distirbución que vale la pena sería de las variables categóricas y de la variable numérica como el tamaño del producto, items, número de visitas por semana y número de delivery. Tal vez se podría hacer una distribución de las coordenadas geográficas, pero no estoy segura.


In [ ]:
# para las columnas categóricas: "category" solo tiene un valor es ques "BEBIDA GASEOSA", Asi que la distribución no tiene mucho sentido.

for i in cat_nominal_columns:
    plt.figure()
    plt.title(f"Distribucion de {i}")
    sns.histplot(data=df, x=i, shrink=0.5)
    plt.xticks(fontsize=8, rotation=45)
    plt.show()


In [ ]:

# items
plt.figure(1)
plt.clf()
plt.title("Distribución de Items")
sns.histplot(data=df, x="items", log_scale=True, bins=20)
plt.yscale("log")
plt.show()

# size
plt.figure(2)
plt.clf()
plt.title("Distribución de tamaño del producto")
sns.histplot(data=df, x="size", log_scale=True, bins=20)
plt.yscale("log")
plt.show()

# num_deliver_per_week

plt.figure(3)
plt.clf()
plt.title("Distribución de Número de entregas por semana")
sns.histplot(data=df, x="num_deliver_per_week", log_scale=False, bins=20)
plt.yscale("log")
plt.show()



In [ ]:
# Acá: transformar ciertas variables categóricas a tipo string

df["customer_id"] = df["customer_id"].astype(str)


In [ ]:
# lo que haré ahora será obtener los top 20 clientes que tienen mayor numero de transacciones y ver que tipo de producto compran, sea por segment, size
# u otra cosa
# y luego ver los que top 20 clientes que tienen el menor número de transacciones.
# lo que viene sería los brands que són más y menos vendidos.
# ver que producto se vende más o tiene más delivery por región

# tal vez hacer un cluster pero ya sería mucho.

### los top 20 clientes que tienen mayor número de transacciones

# para cambiar los top 20 a cualquier número, cambiar 20 por cualquier número

temp_orden_up = df.groupby(["customer_id", "num_deliver_per_week", "X", "Y"])["order_id"].nunique().sort_values(ascending=False)

temp_orden_down = df.groupby("customer_id")["order_id"].nunique().sort_values(ascending=True)


top_up_clientes = temp_orden_up.head(20).reset_index(name="num_transacciones")

top_down_clientes = temp_orden_down.head(20).reset_index(name="num_transacciones")





In [ ]:

# Gráficos: solo para el top 20

plt.figure(4, figsize=(10,8))
plt.clf()
plt.title("Top clientes que registran mayor número de órdenes del producto")
sns.barplot(data=top_up_clientes, x="num_transacciones", y="customer_id", hue="customer_id", palette="viridis")
plt.show()


In [ ]:
# sobre segmentos

segmentos_top = df[df["customer_id"].isin(top_up_clientes["customer_id"])].groupby(["customer_id", 
            "segment"])["order_id"].nunique().reset_index(name="num_transacciones")

# gráfico

plt.figure(5, figsize=(10,8))
plt.clf()
plt.title("Segment según los top clientes")
sns.barplot(data=segmentos_top, x="num_transacciones", y="customer_id", hue="segment", palette="magma")
plt.show()


In [ ]:
# size producto
sizes_top = df[df["customer_id"].isin(top_up_clientes["customer_id"])].groupby(["customer_id", 
            "size"])["order_id"].nunique().reset_index(name="num_transacciones")

sizes_top



In [ ]:
# sobre las brands 

brands_por_zona = df.groupby(["zone_id", "brand"])["order_id"].nunique().reset_index(name="num_transacciones")

brands_top = brands_por_zona.sort_values(["zone_id", "num_transacciones"], ascending=[True, False])


# brands menos vendidas

brands_bottom = brands_por_zona.sort_values(["zone_id", "num_transacciones"], ascending=[True, True])

# gráfico de brands

plt.figure(6, figsize=(10,8))
plt.clf()
plt.title("Top brands que registran mayor número de órdenes en la zona 5148")
sns.barplot(data=brands_top, x="num_transacciones", y="brand", hue="brand", palette="viridis")
plt.xscale("log")
plt.show()



In [ ]:
# sobre las coordenadas X e Y

temp_orden_up = df.groupby(["customer_id", "num_deliver_per_week", "X", "Y", "customer_type"])["order_id"].nunique().sort_values(ascending=False)

temp_orden_down = df.groupby(["customer_id", "num_deliver_per_week", "X", "Y", "customer_type"])["order_id"].nunique().sort_values(ascending=True)


top_up_clientes = temp_orden_up.head(50).reset_index(name="num_transacciones")

top_down_clientes = temp_orden_down.head(50).reset_index(name="num_transacciones")

plt.figure(figsize=(8,6))
sns.scatterplot(data=top_up_clientes, x="X", y="Y", size="num_deliver_per_week", sizes=(20, 400), alpha=0.7)
plt.title("Clientes: ubicación vs entregas y visitas")
plt.legend("")
plt.show()

In [ ]:
plt.figure(figsize=(8,6))
sns.scatterplot(data=top_down_clientes, x="X", y="Y", size="num_deliver_per_week", sizes=(20, 400), alpha=0.7)
plt.title("Clientes: ubicación vs entregas y visitas")
plt.legend("")
plt.show()

In [ ]:
# union de los dataframes top up y top down de clientes, a ver si resulta

import plotly.io as pio
pio.renderers.default = "notebook"

union_top_clientes = pd.concat([top_up_clientes, top_down_clientes])

fig = px.scatter(union_top_clientes, x="X", y="Y", color="customer_type", size="num_deliver_per_week")

fig.show()


In [ ]:
union_top_clientes

## 📌 Holdout [0.25 puntos]

Para evaluar correctamente el modelo y garantizar su capacidad de generalización, se deben dividir los datos en tres conjuntos: 
- `Entrenamiento` : Para ajustar los parámetros.
- `Validación`: Para optimizar hiperparámetros y seleccionar el mejor modelo.
- `Prueba` : Para evaluar el rendimiento final en datos no vistos.

👀 **Hint**: *Recuerde que los datos tienen una temporalidad que debe considerarse al momento de separarlos, para evitar fugas de información. Es importante justificar la estrategia de partición elegida y visualizar la distribución temporal de los conjuntos generados*

In [ ]:
# Desarrolle aquí su código

# la partición de los datos: no tenemos que usar aleatoriedad, si no que trabajamos datos desde cierto un rango temporal y la data de validación en otro rango temporal
# subsiguiente

## 📌 Feature Engineering [0.5 puntos]

<center>
<img src="https://i.imgur.com/CmXZSSC.gif" width="300" height="200">

En esta sección, se deben construir pipelines para automatizar el preprocesamiento de los datos, lo cual garantizará que el flujo de trabajo sea reproducible y eficiente para esta entrega y las futuras. El objetivo es aplicar una serie de transformaciones en un orden definido para asegurar que los datos estén listos para los modelos a entrenar. El pipeline final debe incluir las técnicas de pre-procesamiento que se deben aplicar a los distintos datos (según lo que consideren necesario para el problema). Por ejemplo:

- **Imputación de valores nulos**: Manejo de datos faltantes mediante estrategias adecuadas (media, mediana, moda, interpolación, etc.). 

- **Transformaciones personalizadas**: Uso de ColumnTransformer para aplicar diferentes transformaciones a columnas específicas.

- **Codificación de variables categóricas**: Convertir datos categóricos a un formato numérico adecuado (One-Hot Encoding, Label Encoding, etc.).

- **Discretización de variables**: Conversión de variables numéricas continuas en categorías si son relevantes para el desempeño del modelo a entrenar.

- **Estandarización o normalización** : Ajustar la escala de los datos para mejorar el rendimiento de los algoritmos sensibles a la magnitud de las variables.

- **Eliminación o transformación de valores atípicos**: Identificar y tratar con datos outliers para mejorar la robustez del modelo.

- **Nuevas características** : Creación de variables adicionales que puedan aportar información relevante al modelo.

Cada una de estas transformaciones debe ser justificada en función de su relevancia para el problema y los datos, y es importante evaluar su impacto en el rendimiento del modelo. Además, el pipeline debe ser flexible y modular para poder probar diferentes configuraciones de preprocesamiento.

In [ ]:
# Desarrolle aquí su código 

## 📌 Baseline [0.25 puntos]

<center>
<img src="https://media1.giphy.com/media/v1.Y2lkPTc5MGI3NjExN3lzeGFqZmU3NzJrZHllNjRmaHVzczJpZ29rdHdlMzVpZnQwNXo1diZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/qAtZM2gvjWhPjmclZE/giphy.gif" width="300" height="200">

En esta sección se debe construir el modelo más sencillo posible que pueda resolver el problema planteado, conocido como **Modelo baseline**. Su propósito es servir como referencia para comparar el rendimiento de los modelos más avanzados desarrollados en etapas posteriores.  

Pasos requeridos:  
- Implemente, entrene y evalúe un modelo básico utilizando un pipeline.  
- Asegúrese de incluir en el pipeline las transformaciones del preprocesamiento realizadas previamente junto con un clasificador básico.  
- Evalúe el modelo y presente el informe de métricas utilizando **`classification_report`**.  

Documente claramente cómo se creó el modelo, las decisiones tomadas y los resultados obtenidos. Este modelo será la base comparativa en las secciones posteriores.

In [ ]:
# Desarrolle aquí su código 

## 📌 Elección de modelo [0.75 puntos]

En esta sección deben escoger un modelo que se adapte a las necesidades del negocio. Para esto, pruebe al menos 3 modelos y desarrolle los siguientes aspectos para cada uno:

- **Estructura y diferencias entre los modelos**: Explicar brevemente cada uno y sus hipérparámetros de mayor importancia.
- **Clasificadores recomendados**:
  - `LogisticRegression`
  - `KNeighborsClassifier`
  - `DecisionTreeClassifier`
  - `SVC`
  - `RandomForestClassifier`
  - `LightGBMClassifier` (del paquete `lightgbm`)
  - `XGBClassifier` (del paquete `xgboost`)
  - Otro (según lo que se estime adecuado)
  
- **Evaluación de resultados**: Se utilizará el **`classification_report`** para evaluar el rendimiento de cada modelo, destacando métricas clave como precisión, recall y F1-score. **Importante: No optimicen hiperparámetros, la idea es hacer una selección rápida del modelo.**

**Nota:** Pueden ocupar mas de 1 **instancia** de modelo para resolver el problema (e.g: (modelo_1, grupo_1), (modelo_2, grupo_2), ...).
  
A continuación, se deben responder las siguientes preguntas para evaluar el rendimiento de los modelos entrenados:

1. ¿Hay algún clasificador que supere al modelo baseline?  
2. ¿Cuál es el mejor clasificador entrenado y por qué?  
3. ¿Qué factores hacen que el mejor clasificador sea superior a los otros?  
4. En términos de `tiempo de entrenamiento`, ¿Qué modelo considera más adecuado para experimentar con grillas de optimización?

In [ ]:
# Desarrolle aquí su código 

## 📌 Optimización de Hiperparámetros [1.0 puntos]

<center>
<img src="https://media1.giphy.com/media/v1.Y2lkPTc5MGI3NjExcXJkNzdhYjlneHplaGpsbnVkdzh5dnY3Y2VyaTIzamszdGR1czJ2diZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/2rqEdFfkMzXmo/giphy.gif" width="300" height="200">

A partir de su análisis anterior, se debe proceder a optimizar el rendimiento del modelo seleccionado mediante la optimización de sus hiperparámetros. Para ello, se espera que implementen `Optuna` para optimizar no solo los hiperparámetros del modelo, sino también los de los preprocesadores utilizados (por ejemplo, OneHot Encoding, Scalers, etc.).

Al desarrollar este proceso, deberán responder las siguientes preguntas clave como mínimo:

- ¿Qué métrica decidieron optimizar y por qué?

- ¿Qué hiperparámetro tuvo un mayor impacto en el rendimiento de su modelo?

- ¿Cuánto mejoró el rendimiento del modelo después de la optimización de hiperparámetros?

In [ ]:
# Desarrolle aquí su código

## 📌 Interpretabilidad [1.0 puntos]

En esta sección, deben explicar el funcionamiento de su modelo utilizando las técnicas de interpretabilidad vistas en clase, como `SHAP`. Se espera que sean capaces de descomponer las predicciones y evaluar la importancia de los atributos y las interacciones entre ellos, con el fin de obtener una comprensión más profunda de cómo el modelo toma decisiones. 

Al desarrollar esta parte, deberán responder las siguientes preguntas clave como mínimo:

- ¿Podría explicar el funcionamiento de su modelo para una predicción en particular? Si es así, proporcione al menos tres ejemplos específicos, describiendo cómo el modelo llegó a sus decisiones y qué factores fueron más relevantes en cada caso.

- ¿Qué atributo tiene una mayor importancia en la salida de su modelo? Analice si esto tiene sentido con el problema planteado y justifique la relevancia de dicho atributo en el contexto de las predicciones que se realizan.

- ¿Existe alguna interacción entre atributos que sea relevante para el modelo? Investigue si la combinación de ciertos atributos tiene un impacto significativo en las predicciones y explíquela en **detalle**.

- ¿Podría existir sesgo hacia algún atributo en particular? Reflexione sobre la posibilidad de que el modelo esté favoreciendo ciertos atributos. Si es así, ¿cuál podría ser la causa y qué impacto podría tener esto en la predicción?

In [ ]:
# Desarrolle aquí su código

## 📌 Resultados y Conclusiones [1.0 puntos]

Para finalizar, se deben explicar los desarrollos y resultados obtenidos a lo largo de todo el proceso, desde la selección de las variables hasta la optimización de hiperparámetros e interpretación. Se espera una reflexión crítica sobre el desempeño de los modelos entrenados y una comparación entre los diferentes enfoques. Además, deberán abordar los siguientes puntos clave:

- **Análisis de métricas**: Comenten sobre las métricas obtenidas en cada etapa del modelo, destacando las más relevantes como precisión, recall, F1-score, etc. ¿Cuáles fueron los modelos más efectivos? ¿Qué diferencias notables encontró entre ellos?

- **Impacto de las decisiones tomadas**: Reflexionen sobre cómo las decisiones relacionadas con el preprocesamiento, selección de atributos y optimización de hiperparámetros influyeron en los resultados finales. ¿Hubo alguna decisión que haya tenido un impacto notable en el rendimiento?

- **Lecciones aprendidas**: Concluyan sobre las lecciones más importantes que aprendieron durante el proceso y cómo estas pueden influir en futuras iteraciones del modelo. ¿Qué se podría mejorar si se repitiera el proceso? Si tuvieran más recursos y tiempo, ¿qué otras técnicas/herramientas habrían utilizado?

> [Escriba aquí sus resultados]

Mucho éxito!

<center>
<img src="https://media3.giphy.com/media/v1.Y2lkPTc5MGI3NjExaHpvOTY5Z3hpdHI3aDBpdGRueXRqamZncXp2emFrbjJ5M2s5eTR1dSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/1PMVNNKVIL8Ig/giphy.gif" width="300" height="200">
